dex/src/main/scala/com/wavesplatform/dex/Matcher.scala

```diff
@@ -82,8 +88,44 @@ class Matcher(settings: MatcherSettings, gRPCExtensionClient: DEXClient)(implici
                                                                   wavesBlockchainAsyncClient.hasScript,
                                                                   wavesBlockchainAsyncClient.isFeatureActivated)
 
-  private val assetDecimalsCache                           = new AssetDecimalsCache(wavesBlockchainSyncClient.assetDescription)
-  private implicit val errorContext: ErrorFormatterContext = (asset: Asset) => assetDecimalsCache.get(asset)
+  private val waves = liftValueAsync[AssetsDB.Item](
+    AssetsDB.Item(
+      name = ByteStr(AssetPair.WavesName.getBytes(StandardCharsets.UTF_8)),
+      decimals = 8
+    ))
+  private def getDecimals(assetsDB: AssetsDB, assetDesc: IssuedAsset => Future[Option[BriefAssetDescription]])(asset: Asset)(
+      implicit ec: ExecutionContext): EitherT[Future, MatcherError, (Asset, Int)] = getDesc(assetsDB, assetDesc)(asset).map(x => asset -> x.decimals)
+
+  private def getDesc(assetsDB: AssetsDB, assetDesc: IssuedAsset => Future[Option[BriefAssetDescription]])(asset: Asset)(
+      implicit ec: ExecutionContext): EitherT[Future, MatcherError, AssetsDB.Item] = {
+    log.info(s"getDesc(${AssetPair.assetIdStr(asset)})")
+    asset match {
+      case Waves => waves
+      case asset: IssuedAsset =>
+        assetsDB.get(asset) match {
+          case Some(x) => liftValueAsync[AssetsDB.Item](x)
+          case None =>
+            EitherT {
+              assetDesc(asset)
+                .map {
+                  _.toRight[MatcherError](error.AssetNotFound(asset))
+                    .map { desc => // TODO andThen
+                      log.info(s"Got description for ${AssetPair.assetIdStr(asset)}: $desc")
+                      val r = AssetsDB.Item(desc.name, desc.decimals)
+                      assetsDB.put(asset, r)
+                      r
+                    }
+                }
+            }
+        }
+    }
+  }
+
+  private lazy val assetsDB: AssetsDB = AssetsDB(db)
```

Comment:
What do you think about dedicated class that will incapsulate this logic inside?
Matcher becomes trash class

Reply:
Moved to the companion object (changes on the way :) )